### ISSUE TITLE
[HOLD for payment 2024-10-09] [$250] Onboarding - Onboarding Modal opens when returning back online

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#49499]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.38-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Sign in with a new user
2. On the Onboarding modal turn off the internet
3. Go through the onboarding flow and click continue
4. Refresh the page to kill the app
5. Go back online and reload

## Expected Result:
Reloading after going back online should not bring the Onboarding flow back

## Actual Result:
Reloading after going back online bring back the Onboarding flow for a moment

## Workaround:
Unknown

## Platforms:

- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/29ca040f-6dff-4547-8102-88ecf81324e0

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021836961456209861456</li>
        <li>Upwork Job ID: 1836961456209861456</li>
        <li>Last Price Increase: 2024-09-20</li>
<li>Automatic offers: </li>
<ul>
<li>allgandalf | Contributor | 104080039</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @zanyrenney</details>

**Comment by IuliiaHerets** [comment]
@jliexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by jliexpensify** [comment]
Able to repro this one

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Onboarding Modal opens when returning back online

### What is the root cause of that problem?

`CompleteGuidedSetup` will be stored locally when there is no internet connection, and `hasCompletedGuidedSetupFlow` will not be updated because `CompleteGuidedSetup` has not been executed yet. Once the connection is restored, the condition checking `hasCompletedGuidedSetupFlow` will still return `false` until the `CompleteGuidedSetup` API is fully executed. As a result, the onboarding process will reappear because the condition to show onboarding is checked before `CompleteGuidedSetup` is completed.

### What changes do you think we should make in order to solve the problem?
We need to add a condition for displaying onboarding. If there is a `CompleteGuidedSetup` pending for the account, onboarding will not be displayed. By using this method, we can avoid the scenario where, if the `CompleteGuidedSetup` API call encounters an error, onboarding will still follow the correct flow

```diff
function hasCompletedGuidedSetupFlowSelector(onboarding: OnyxValue<typeof ONYXKEYS.NVP_ONBOARDING>): boolean {
    // onboarding is an array for old accounts and accounts created from olddot
    if (Array.isArray(onboarding)) {
        return true;
    }

+    const persistedRequests = PersistedRequests.getAll();
+    const hasPendingCompleteGuidedSetupRequest = persistedRequests.some((item) => item.command === WRITE_COMMANDS.COMPLETE_GUIDED_SETUP);

+    // If there's a pending request to complete guided setup, return false
+    if (hasPendingCompleteGuidedSetupRequest) {
+        return false;
+    }

    return onboarding?.hasCompletedGuidedSetupFlow ?? true;
}
```

<summary>
<details>POC

https://github.com/user-attachments/assets/d62f32b1-07b7-41de-868c-aff6c132e975

</details>
</summary>

**Comment by CyberAndrii** [comment]
This is a regression from https://github.com/Expensify/App/pull/49263 and was fixed in https://github.com/Expensify/App/pull/49516 by removing the `hasCompletedGuidedSetupFlowSelector` call. Ideally we should still revert https://github.com/Expensify/App/pull/49263 to prevent bugs in other places

**Comment by jliexpensify** [comment]
Thanks @CyberAndrii - have assigned Vit and @allgandalf to have a look at this and confirm it's a regression fornm their PR.

**Comment by mountiny** [comment]
I think this is minor, the original bug the Pr was fixing was much more critical

@blazejkustra would you be able to follow up on this one and see why the modal is opening for a bit? Thanks!

**Comment by CyberAndrii** [comment]
To clarify, this issue was already fixed by https://github.com/Expensify/App/pull/49516. It's just that https://github.com/Expensify/App/pull/49263 changed the default value in the `hasCompletedGuidedSetupFlowSelector` function from `false` to `true`. Then the follow-up PR fixed the initial issue using a completely different appoach (to avoid bugs like this) that no longer uses `hasCompletedGuidedSetupFlowSelector`. And since it's no longer used, it would make sense to revert the default value to what it was previously.

**Comment by huult** [comment]
We can use [my proposal](https://github.com/Expensify/App/issues/49499#issuecomment-2363079211) to fix this issue because it solves the problem without affecting other flows.

**Comment by blazejkustra** [comment]
Update: I have a working fix for this [here](https://github.com/Expensify/App/pull/49586), but I'm struggling to make the exit animation of the Modal to work correctly, I'm sure I'll update on the progress rather quickly

**Comment by blazejkustra** [comment]
>  I'm sure I'll update on the progress rather quickly

Famous last words 💀 

Update: We investigated this with @adamgrzybowski, as it is deeply connected to navigation logic, there are several issues with the current setup of onboarding (redundant logic, weird flows etc)
- `hasCompletedGuidedSetupFlow` is not optimistically updated when onboarding is finished
- onboardingFlow is initiated from `BottomTabBar.tsx` component (❓❗️) 
- For some reason `NAVIGATORS.ONBOARDING_MODAL_NAVIGATOR` is always mounted, while nested screens are not always accessible
- There is a sophisticated logic inside `OnboardingModalNavigator.tsx` to navigate back once the onboarding is finished

All that being said we come up with a solution that refactors a lot of code, and could potentially affect HybridApp - that's why first thing in the morning I'll check with the HybridApp team to make sure we don't introduce any new bugs.

cc @mountiny @allgandalf

**Comment by mountiny** [comment]
@blazejkustra thanks for the update, how its going?

**Comment by blazejkustra** [comment]
I'll update shortly, we are testing the refactor on HybridApp with @mateuuszzzzz

**Comment by blazejkustra** [comment]
Update: I added an explanation to the draft PR [here](https://github.com/Expensify/App/pull/49586)

After a long discussion with both @adamgrzybowski (navigation) and @mateuuszzzzz (hybrid), we think the onboarding flow is unnecessary complicated and some of the logic is redundant. This makes all work connected to the onboarding very tedious and full of edge cases (especially on HybridApp).

I think we should refactor the onboarding logic as part of this PR, because it's not trivial how to fix the offline behaviour without breaking new things. You can check the draft PR to see the necessary changes, we are also expecting some changes in the hybrid app as we discovered some hidden bugs.

Thoughts? @mountiny @allgandalf @jliexpensify

**Comment by blazejkustra** [comment]
Alternative approach would be to try to fix this issue without the big refactor, but as mentioned above it could prove complicated due to `OnboardingNavigator` conditional rendering (I have a working approach but the onboarding modal disappear animation is cut off)

**Comment by blazejkustra** [comment]
Do you think we should proceed with the refactor? @mountiny @jliexpensify

**Comment by blazejkustra** [comment]
Yes, but it's not "complete" onboarding refactor, as the onboarding modals are the same - just the logic is brought together ino one place + it's simplified.

**Comment by mountiny** [comment]
@blazejkustra lets do it right and properly so if that requires a bigger refactor, lets do that

**Comment by blazejkustra** [comment]
Awesome, I'm excited to do it right 🤩 @allgandalf mind taking a first look at the [PR](https://github.com/Expensify/App/pull/49586)?

**Comment by jliexpensify** [comment]
Not overdue!

**Comment by jliexpensify** [comment]
Just a heads up I'm OOO from 3rd to 14th, but I don't think anything is needed from me during this period. Feel free to reassign to another B0 if needed - otherwise I will catch up when I'm back.

**Comment by mountiny** [comment]
$250 to @allgandalf for their review

**Comment by mountiny** [comment]
Yep I think this is fair, thanks for your willingness to step up and help with this one
